File: solutions/feb_2024/1481_least_number_of_unique_integers_after_k_removals/solution.py
from collections import Counter
from typing import List


class Solution:
    def findLeastNumOfUniqueInts(self, arr: List[int], k: int) -> int:
        counter = Counter(arr)
        frequencies = sorted(counter.items(), key=lambda item: item[1])
        removed_count = 0

        for _, freq in frequencies:
            if k >= freq:
                k -= freq
                removed_count += 1
            else:
                break

        return len(frequencies) - removed_count

File: solutions/feb_2024/1481_least_number_of_unique_integers_after_k_removals/solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
  public int findLeastNumOfUniqueInts(int[] arr, int k) {
    Map<Integer, Integer> counter = new HashMap<>();
    for (int num : arr) {
      counter.put(num, counter.getOrDefault(num, 0) + 1);
    }

    List<Integer> frequencies = new ArrayList<>(counter.values());
    Collections.sort(frequencies);

    int removedCount = 0;
    for (int freq : frequencies) {
      if (k < freq) {
        break;
      }

      k -= freq;
      removedCount += 1;
    }

    return frequencies.size() - removedCount;
  }
}
